fdbserver/DiskQueue.actor.cpp

```diff
@@ -270,68 +270,103 @@ class RawDiskQueue_TwoFiles : public Tracked<RawDiskQueue_TwoFiles> {
 
 	Future<Void> truncateFile(int file, int64_t pos) { return truncateFile(this, file, pos); }
 
-	Future<Void> push(StringRef pageData, vector<Reference<SyncQueue>>& toSync) {
+	// FIXME: Merge this function with IAsyncFileSystem::incrementalDeleteFile().
+	ACTOR static void incrementalTruncate(Reference<IAsyncFile> file) {
+		state int64_t remainingFileSize = wait( file->size() );
+
+		for( ; remainingFileSize > 0; remainingFileSize -= FLOW_KNOBS->INCREMENTAL_DELETE_TRUNCATE_AMOUNT ){
+			wait(file->truncate(remainingFileSize));
+			wait(file->sync());
+			wait(delay(FLOW_KNOBS->INCREMENTAL_DELETE_INTERVAL));
+		}
+	}
+
+	ACTOR static Future<Reference<IAsyncFile>> replaceFile(Reference<IAsyncFile> toReplace) {
+		incrementalTruncate( toReplace );
+
+		Reference<IAsyncFile> _replacement = wait( IAsyncFileSystem::filesystem()->open( toReplace->getFilename(), IAsyncFile::OPEN_ATOMIC_WRITE_AND_CREATE | IAsyncFile::OPEN_READWRITE | IAsyncFile::OPEN_UNCACHED | IAsyncFile::OPEN_UNBUFFERED | IAsyncFile::OPEN_LOCK, 0 ) );
+		state Reference<IAsyncFile> replacement = _replacement;
+		wait( replacement->sync() );
+
+		return replacement;
+	}
+
+	Future<Void> push(StringRef pageData, vector<Reference<SyncQueue>>* toSync) {
+		return push( this, pageData, toSync );
+	}
+
+	ACTOR static Future<Void> push(RawDiskQueue_TwoFiles* self, StringRef pageData, vector<Reference<SyncQueue>>* toSync) {
```

Comment:
Given that we have infrastructure like `Reference` in place, maybe consider use that instead of raw pointers for the sake of consistency and safety.

Reply:
In my head I view Reference<> as approximately equal to std::shared_ptr<>, so using it here feels like it would wrongly indicate the ownership semantics.  We have a non-owning pointer to an object that has a clear lifetime in our parent scope.  There's an std::experimental::observer_ptr<> that I could rewrite it to, but that seems like overkill.

Reply:
It seems to me that it does need a owning pointer (or a `mutable borrow` in `Rust`) because `push` is indeed mutating state of `self`. Or are we talking different owning semantics?

Reply:
Never mind I just read the documentation of `observer_ptr` and now understand what you meant by `non-owning`.